### ISSUE TITLE
Run on remote clusters

### ISSUE THREAD

**Issue by nirs** [nirs/kubectl-gather#13]
When working with local clusters like a testing environment collecting data is fast. But when working with remote clusters the network round trip and bandwidth limits can make collection too slow. In this case it can be useful to gather the data on the remote cluster and download the gathered data at the end.

We can use oc adm must-gather for running kubectl-gather on the remote cluster, by creating a gather container image using must-gather image API:
https://github.com/openshift/enhancements/blob/master/enhancements/oc/must-gather.md#must-gather-images

To support multiple clusters, we can create an option for running remotely, running multiple `oc adm must-gather` in parallel with multiple clusters, storing the data in the same directory.

## Example usage

Possible user interface

    kubectl gather -d gather --contexts hub,dr1,dr2 --remote

This will run in parallel:

    oc adm must-gather --image=registry/name/gather:tag --context hub --dest-dir gather/hub
    oc adm must-gather --image=registry/name/gather:tag --context dr1 --dest-dir gather/dr1
    oc adm must-gather --image=registry/name/gather:tag --context dr2 --dest-dir gather/dr2

To provide arguments we can use:

    oc adm must-gather --image=registry/name/gather:tag \
        --context cluster \
        --namespaces foo,bar \
        --dest-dir gather/cluster
        -- /usr/bin/gather -n foo,bar

To get consistent output we can gather data into temporary directory, and move the contents of registry-gather-sha256-6835423dbaa76467b279bc91d9aad3e9c9022eebe9dd7c7e37f6ead8f2c0ed9f/ into the final location:

    oc adm must-gather --image=registry/name/gather:tag --context dr2 --dest-dir gather/dr2/.tmp
    mv gather/dr2/.tmp/registry-gather-sha256-*/* gather/dr2
    rm -rf gather/dr2/.tmp

## Running in cluster

When running inside a cluster some things need to change:

- We must collect data in /must-gather
- There is not kubeconfig - the pod can connect to the api server using the serviceacount mount:
  - https://kubernetes.io/docs/tasks/run-application/access-api-from-pod/
  - https://github.com/kubernetes/client-go/blob/master/examples/in-cluster-client-configuration/main.go
- `--contexts` cannot be used since there is no kube config
- We cannot not create the context directory since we don't have a context
- We need additional executable in the image to get logs (tar, kubectl)
  - kubectl can be eliminated in we implement remote commands using [client-go/tools/remotecommand](https://pkg.go.dev/k8s.io/client-go/tools/remotecommand)
- Need rsync to allow must-gather to copy the output efficiently

## Future work

- Implement remote commands without kubectl so we can remove kubectl from the image
